src/plugins/wizard/public/application/components/top_nav.tsx

```diff
@@ -5,14 +5,29 @@
 
 import React, { useMemo } from 'react';
 import { useParams } from 'react-router-dom';
+import useUnmount from 'react-use/lib/useUnmount';
 import { PLUGIN_ID } from '../../../common';
 import { useOpenSearchDashboards } from '../../../../opensearch_dashboards_react/public';
 import { getTopNavConfig } from '../utils/get_top_nav_config';
 import { WizardServices } from '../../types';
 
 import './top_nav.scss';
 import { useIndexPatterns, useSavedWizardVis } from '../utils/use';
-import { useTypedSelector } from '../utils/state_management';
+import { useTypedSelector, useTypedDispatch } from '../utils/state_management';
+import { setSaveState } from '../utils/state_management/metadata_slice';
+
+// TODO: Need to finalize the error messages
+const saveButtonMsg = (isEmpty, hasChange, hasUnappliedChanges) => {
```

Comment:
Nit: You can abstract all this save logic into a separate hook like `useCanSave`.

e.g.

```js
const useCanSave = () => {
  let errorMsg = ''
  const isEmpty = useTypedSelector((state) => 
    state.visualization.activeVisualization?.aggConfigParams?.length === 0
  ); 
  // Other selectors and errorMsg logic

  return {
    canSave: !isEmpty && hasChange && !hasUnappliedChanges,
    errorMsg
  }
}
```

This will abstract away the whole `canSave` logic that you can independently write tests for 😄

Reply:
I wrote a separate hook and return an object that includes the canSave boolean and errorMsg string. However, i don't think we can include those useTypedSelector hooks inside the separate hook because i think they are supposed to place in the top level function which is TopNav(). If i place them in a separate hook, it will prompt errors.

Reply:
I dont think thats true. We are doing it here in [`useVisualizationType` ](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/wizard/public/application/utils/use/use_visualization_type.ts)

Reply:
yea i misunderstood it first. Changed in the newest commit.